File: src/atla_insights/metadata.py
"""Utils for the atla_insights package."""

import json
import logging
from itertools import islice
from typing import Optional

from atla_insights.constants import (
    MAX_METADATA_FIELDS,
    MAX_METADATA_KEY_CHARS,
    MAX_METADATA_VALUE_CHARS,
    METADATA_MARK,
    OTEL_MODULE_NAME,
)
from atla_insights.context import metadata_var, root_span_var

logger = logging.getLogger(OTEL_MODULE_NAME)


def _truncate_value(value: str, max_chars: int) -> str:
    """Truncate a value to a maximum number of characters.

    :param value (str): The value to truncate.
    :param max_chars (int): The maximum number of characters to allow.
    :return (str): The truncated value.
    """
    if len(value) > max_chars:
        return value[: (max_chars - 3)] + "..."
    return value


def validate_metadata(metadata: dict[str, str]) -> dict[str, str]:
    """Validate the user-provided metadata field.

    :param metadata (dict[str, str]): The metadata field to validate.
    :return (dict[str, str]): The validated metadata.
    """
    if not isinstance(metadata, dict):
        raise ValueError("The metadata field must be a dictionary.")

    if not all(isinstance(k, str) and isinstance(v, str) for k, v in metadata.items()):
        logger.error("The metadata field must be a mapping of string to string.")
        metadata = {str(k): str(v) for k, v in metadata.items()}

    if len(metadata) > MAX_METADATA_FIELDS:
        logger.error(
            f"The metadata field has {len(metadata)} fields, "
            f"but the maximum is {MAX_METADATA_FIELDS}."
        )
        metadata = dict(islice(metadata.items(), MAX_METADATA_FIELDS))

    if any(len(k) > MAX_METADATA_KEY_CHARS for k in metadata.keys()):
        logger.error(
            "The metadata field must have keys with less than "
            f"{MAX_METADATA_KEY_CHARS} characters."
        )
        metadata = {
            _truncate_value(k, MAX_METADATA_KEY_CHARS): v for k, v in metadata.items()
        }

    if any(len(v) > MAX_METADATA_VALUE_CHARS for v in metadata.values()):
        logger.error(
            "The metadata field must have values with less than "
            f"{MAX_METADATA_VALUE_CHARS} characters."
        )
        metadata = {
            k: _truncate_value(v, MAX_METADATA_VALUE_CHARS) for k, v in metadata.items()
        }

    return metadata


def get_metadata() -> Optional[dict[str, str]]:
    """Get the metadata for the current trace.

    :return (Optional[dict[str, str]]): The metadata for the current trace.
    """
    return metadata_var.get()


def set_metadata(metadata: dict[str, str]) -> None:
    """Set the metadata for the current trace.

    ```py
    from atla_insights import instrument, set_metadata

    @instrument("My Function")
    def my_function():
        set_metadata({"some_key": "some_value", "other_key": "other_value"})
        ...
    ```

    :param metadata (dict[str, str]): The metadata to set for the current trace.
    """
    metadata = validate_metadata(metadata)

    metadata_var.set(metadata)
    if root_span := root_span_var.get():
        # If the root span already exists, we can assign the metadata to it.
        # If not, it will be assigned the `metadata_var` context var on creation.
        root_span.set_attribute(METADATA_MARK, json.dumps(metadata))

File: js/src/metadata.ts
/**
 * Metadata management for Atla Insights.
 * Works with both configure() and instrumentVercel() approaches.
 */
import { trace, context as otelContext } from "@opentelemetry/api";
import { METADATA_MARK } from "./internal/constants";

const METADATA_CONTEXT_KEY = Symbol("atla.metadata");

// Metadata validation limits (matching Python SDK)
const MAX_METADATA_FIELDS = 25;
const MAX_METADATA_KEY_CHARS = 40;
const MAX_METADATA_VALUE_CHARS = 100;

/**
 * Truncate a string value to the specified maximum length.
 */
function truncateValue(value: string, maxLength: number): string {
    return value.length > maxLength ? value.substring(0, maxLength) : value;
}

/**
 * Validate the user-provided metadata field.
 * Mirrors the Python SDK's _validate_metadata function.
 */
function validateMetadata(metadata: Record<string, string>): Record<string, string> {
    if (typeof metadata !== 'object' || metadata === null || Array.isArray(metadata)) {
        throw new Error("The metadata field must be a dictionary.");
    }

    // Verify all keys and values are strings
    for (const [key, value] of Object.entries(metadata)) {
        if (typeof key !== 'string' || typeof value !== 'string') {
            console.error("The metadata field must be a mapping of string to string.");
            // Convert non-strings to strings as fallback
            const stringKey = String(key);
            const stringValue = String(value);
            metadata = { ...metadata };
            delete metadata[key];
            metadata[stringKey] = stringValue;
        }
    }

    // Limit number of fields
    if (Object.keys(metadata).length > MAX_METADATA_FIELDS) {
        console.error(
            `The metadata field has ${Object.keys(metadata).length} fields, ` +
            `but the maximum is ${MAX_METADATA_FIELDS}.`
        );
        const entries = Object.entries(metadata).slice(0, MAX_METADATA_FIELDS);
        metadata = Object.fromEntries(entries);
    }

    // Truncate oversized keys
    const oversizedKeys = Object.keys(metadata).filter(k => k.length > MAX_METADATA_KEY_CHARS);
    if (oversizedKeys.length > 0) {
        console.error(
            `The metadata field must have keys with less than ${MAX_METADATA_KEY_CHARS} characters.`
        );
        const newMetadata: Record<string, string> = {};
        for (const [key, value] of Object.entries(metadata)) {
            const truncatedKey = truncateValue(key, MAX_METADATA_KEY_CHARS);
            newMetadata[truncatedKey] = value;
        }
        metadata = newMetadata;
    }

    // Truncate oversized values
    const oversizedValues = Object.values(metadata).filter(v => v.length > MAX_METADATA_VALUE_CHARS);
    if (oversizedValues.length > 0) {
        console.error(
            `The metadata field must have values with less than ${MAX_METADATA_VALUE_CHARS} characters.`
        );
        for (const [key, value] of Object.entries(metadata)) {
            if (value.length > MAX_METADATA_VALUE_CHARS) {
                metadata[key] = truncateValue(value, MAX_METADATA_VALUE_CHARS);
            }
        }
    }

    return metadata;
}

/**
 * Set metadata that will be added to all spans in the current trace.
 * This is for runtime metadata updates within a trace.
 */
export function setMetadata(metadata: Record<string, string>): void {
    const span = trace.getActiveSpan();
    if (span) {
        const validatedMetadata = validateMetadata(metadata);

        // Get existing metadata from the span if any
        const existingMetadata = span.attributes[METADATA_MARK];
        let currentMetadata: Record<string, string> = {};

        if (existingMetadata && typeof existingMetadata === 'string') {
            try {
                currentMetadata = JSON.parse(existingMetadata);
            } catch (e) {
                // Ignore parse errors
            }
        }

        // Merge with new metadata
        const mergedMetadata = { ...currentMetadata, ...validatedMetadata };
        span.setAttribute(METADATA_MARK, JSON.stringify(mergedMetadata));
    }
}

/**
 * Get current metadata from the active span.
 */
export function getMetadata(): Record<string, string> | undefined {
    const span = trace.getActiveSpan();
    if (span) {
        const metadataAttr = span.attributes[METADATA_MARK];
        if (metadataAttr && typeof metadataAttr === 'string') {
            try {
                return JSON.parse(metadataAttr);
            } catch (e) {
                // Ignore parse errors
            }
        }
    }
    return undefined;
}

/**
 * Run a function with additional metadata in context.
 */
export function withMetadata<T>(
    metadata: Record<string, string>,
    fn: () => T | Promise<T>
): T | Promise<T> {
    const validatedMetadata = validateMetadata(metadata);
    const ctx = otelContext.active();
    const currentMetadata = ctx.getValue(METADATA_CONTEXT_KEY) as Record<string, string> || {};
    const newCtx = ctx.setValue(METADATA_CONTEXT_KEY, { ...currentMetadata, ...validatedMetadata });

    return otelContext.with(newCtx, fn);
}

/**
 * Clear all metadata from the active span.
 */
export function clearMetadata(): void {
    const span = trace.getActiveSpan();
    if (span) {
        span.setAttribute(METADATA_MARK, "{}");
    }
}
